### PR TITLE
Issue #1161 - Use correct resource version in K8S watch API calls to avoid lost update events

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/1161

Following changes are implemented:

* `ctl.WatchResources` method now supply `resourceVersion` parameter as described in https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes

* `ctl.GetResources` return channel of lists (one per group kind) instead of one giant list of everything. This allows controller to process one list and free memory sooner

### CODE REVIEW

File: controller/cache/cache.go [hunk 1]
```diff
@@ -191,9 +192,12 @@ func (c *liveStateCache) Run(ctx context.Context) {
 			// cluster resources must be watched only if cluster has at least one app
 			if (event.Type == watch.Deleted || !hasApps) && ok {
 				info.cancel()
+				watchingClustersLock.Lock()
 				delete(watchingClusters, event.Cluster.Server)
+				watchingClustersLock.Unlock()
```
Comment (alexec) [inline controller/cache/cache.go]: should this be on defer?

File: controller/cache/cache.go [hunk 2]
```diff
@@ -204,6 +208,7 @@ func (c *liveStateCache) Run(ctx context.Context) {
 					},
 					cluster: event.Cluster,
 				}
+				watchingClustersLock.Unlock()
```
Comment (alexec) [inline controller/cache/cache.go]: defer?

File: controller/cache/cluster.go [hunk 3]
```diff
@@ -21,12 +21,17 @@ import (
 )
 
 const (
-	clusterSyncTimeout  = 1 * time.Hour
+	clusterSyncTimeout  = 24 * time.Hour
```
Comment (alexec) [inline controller/cache/cluster.go]: accidental?

**Review by alexec — COMMENTED** [review]
Can we pair on the review of this so I can get a better understanding? Or is it better @jessesuen reviews first?

File: controller/cache/cache.go (hunk 1 above)
Comment (alexmt) [reply controller/cache/cache.go]: to use defer I would have to create one line function, so I decided to just execute unlock manually.

File: controller/cache/cluster.go (hunk 3 above)
Comment (alexmt) [reply controller/cache/cluster.go]: This is intentional. I want re-sync cluster cache less often since now controller should never miss watch events.

**Review by jessesuen — APPROVED** [review]

**Pull request merged**